Add playbook entry for sample with repeated timestamp (#157)
grafana/cortex-jsonnet: cortex-mixin/docs/playbooks.md

- `Retain`: the volume will not be deleted until the PV resource will be manually deleted from Kubernetes
- `Delete`: the volume will be automatically deleted

## Log lines

### Log line containing 'sample with repeated timestamp but different value'

This means a sample with the same timestamp as the latest one was received with a different value. The number of occurrences is recorded in the `cortex_discarded_samples_total` metric with the label `reason="new-value-for-timestamp"`.

Possible reasons for this are:

> - Multiple Prometheus servers / Grafana agents are scraping the same target without deduplication in place. Check the IP addresses mentioned in the log line for the agent that returned the duplicate sample. Change the labels of each sample generated per agent so they are unique.

Reviewer: Few things, please:
1. As I mentioned in my previous comment https://github.com/grafana/cortex-jsonnet/pull/157#discussion_r467106893 I'm unsure this could be the cause for this specific error. Timestamps are different when the same target is monitored by different Prometheus servers. The only case could happen is when the monitored target expose metrics with timestamps **and** these timestamps don't change between the scraping done by two different Prometheus servers (eg. may happen with cAdvisor metrics).
2. Whenever you refer to "agent" only you should always mention "Prometheus / Agent" because the Agent is a less common case than running Prometheus and this is a public playbook

Author: Yeah, you're right. I spoke with Goutham about this specific problem which one customer was having. They had HA pairs setup that were not correctly using deduplication. But looking at my notes that might have been another issue. I'll remove this reason.

> - Incorrect relabelling rules can cause a label to be dropped from a sample so that multiple samples have the same labels. If these samples were collected from the same target they will have the same timestamp. An example is dropping the `cpu` label when there are multiple cpus.

Reviewer: ```suggestion
- Incorrect relabelling rules can cause a label to be dropped from a series so that multiple series have the same labels. If these series were collected from the same target they will have the same timestamp.
```

- `Retain`: the volume will not be deleted until the PV resource will be manually deleted from Kubernetes
- `Delete`: the volume will be automatically deleted

## Log lines

### Log line containing 'sample with repeated timestamp but different value'

> This means a sample with the same timestamp as an existing one was received with a different value. The number of occurrences is recorded in the `prometheus_target_scrapes_sample_out_of_order_total` metric.

Reviewer: ```suggestion
This means a sample with the same timestamp as the latest one was received with a different value. The number of occurrences is recorded in the `cortex_discarded_samples_total` metric with the label `reason="new-value-for-timestamp"`.
```

Possible reasons for this are:

> - Multiple agents are scraping the same app without deduplication in place. Check the IP addresses mentioned in the log line for the agent that returned the deplicate sample. Change the labels of each sample generated per agent so they are unique.

Reviewer 2: ```suggestion
- Multiple agents are scraping the same app without deduplication in place. Check the IP addresses mentioned in the log line for the agent that returned the duplicate sample. Change the labels of each sample generated per agent so they are unique.
```

> - Multiple agents are scraping the same app without deduplication in place. Check the IP addresses mentioned in the log line for the agent that returned the duplicate sample. Change the labels of each sample generated per agent so they are unique.

Reviewer: More thoughts about it:
- Unlikely to be the cause. Timestamp are generally different when two independent processes scrape the same target. Timestamps could be the same when they're included in the `/metrics` response, but it's very uncommon (eg. among common exporters, I'm aware only cAdvisor doing it)
- Please always refer to Prometheus too as remote write source (it's more common than the Grafana agent)
- "Change the labels of each sample generated per agent" do you mean add an external label to the agent to make its series unique? Ok, but we shouldn't have multiple agents scraping the same targets unless they're configured as HA pairs in Cortex (see https://cortexmetrics.io/docs/production/ha-pair-handling/)

```suggestion
- Multiple Prometheus servers / Grafana agents are scraping the same target without deduplication in place. Check the IP addresses mentioned in the log line for the agent that returned the duplicate sample. Change the labels of each sample generated per agent so they are unique.
```

> - Incorrect relabelling rules can cause a label to be dropped from a sample so that multiple samples have the same labels. If these samples were collected at the same time they will cause this error.

Reviewer: > If these samples were collected at the same time they will cause this error.

In other terms, this is generally caused when the clashing series are coming from the same exact target. So let's say a target expose these metrics:

```
container_cpu_usage_seconds{cpu="0"} 123
container_cpu_usage_seconds{cpu="1"} 456
```

Then you have a relabelling rule removing the cpu and you end up with clashing series with the same exact timestamp (because both series are scraped within the same scraping activity):
```
container_cpu_usage_seconds 123
container_cpu_usage_seconds 456
```

- The exporter being scraped sets the same timestamp on every scrape. Note that exporters should generally not set timestamps.

> - Prometheus scrapes at the millisecond level. If the scrapes are done very quickly the same sample could be returned. This is very unlikely.

Reviewer: I would remove this. Not a real case.
